src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,49 +1,111 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int MAX_CAPACITY = Integer.MAX_VALUE - 8;
+    private static final int DEFAULT_CAPACITY = 10;
+    private T[] elementData;
+    private int size;
+    private int capacity;
+
+    public ArrayList() {
+        elementData = (T[]) new Object[DEFAULT_CAPACITY];
+        capacity = DEFAULT_CAPACITY;
+    }
 
     @Override
     public void add(T value) {
-
+        ensureCapacity(size + 1);
+        elementData[size++] = value;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index == size) {
+            add(value);
+            return;
+        }
+        checkIndex(index);
+        ensureCapacity(size + 1);
+        System.arraycopy(elementData, index, elementData, index + 1, size - index);
+        elementData[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        if (list == null) {
+            return;
+        }
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return elementData[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkIndex(index);
+        elementData[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        T value = elementData[index];
+        System.arraycopy(elementData,index + 1, elementData, index, size - index - 1);
+        elementData[size--] = null;
+        return value;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        for (int i = 0; i < size; i++) {
+            if (element == elementData[i] || elementData[i] != null
+                    && elementData[i].equals(element)) {
+                T value = elementData[i];
+                System.arraycopy(elementData,i + 1, elementData, i, size - i - 1);
+                elementData[size--] = null;
+                return value;
```

Comment:
can we call method remove by index here?

Reply:
done